Fix: Checkout Issue in self hosted runner due to faulty submodule check-ins

megamanics opened the pull request:
Fixes issue related to faulty submodule checkins - this affects the self hosted runners.
Github hosted runner cleanup every time so this is not a problem when using Github hosted runners.

### Without fix - Self hosted runners fail due to work directory corruption.
![image](https://user-images.githubusercontent.com/10250297/222324893-eb121693-2d00-403a-b319-ba517a386c03.png)

### With Fix
![image](https://user-images.githubusercontent.com/10250297/222485458-ee17715d-0253-443e-a1a1-5f660e490d87.png)

Checks for bad submodule within the directory for self hosted runners and sets flag for cleanup.
```
git submodule status; echo $?
fatal: no submodule mapping found in .gitmodules for path 'submodule-repo'
128
```

### Fixes
Resolves https://github.com/github/c2c-actions-support/issues/1511
Resolves https://github.zendesk.com/agent/tickets/1821873
Resolves https://github.com/actions/checkout/issues/933
Maybe related: https://github.com/actions/checkout/issues/1148

megamanics committed:
Fix Self hosted runner issue wrt bad submodules - solution cleanup working space.

megamanics committed:
Fix format with npm run format output

megamanics committed:
Add mock implementation for new function submoduleStatus

sonialee3 reviewed (commented):
Thx

sonialee3 commented:
Thx

carlospontual commented:
I've been using it on some of my projects that had the submodules issue, and I can say that it really fixed the issues we were facing in multiple of our branches.

Thank you @megamanics for your work (I'm using action `megamanics/checkout@d12863bffd0b89f66bb35c44a231bb2fdc5e3dad`). I hope this PR gets approved, merged & released asap as part of the latest `actions/checkout@v3` 👍🏼 🏆

Link- reviewed (changes requested):
Thank you for working on this fix 🙏 - we need to add some tests to this change. I also want someone else from the team to give this another look.

Link- committed:
Merge branch 'main' into megamanics/main

megamanics commented on __test__/git-directory-helper.test.ts:
```diff
@@ -423,6 +423,9 @@ async function setup(testName: string): Promise<void> {
     submoduleForeach: jest.fn(),
     submoduleSync: jest.fn(),
     submoduleUpdate: jest.fn(),
+    submoduleStatus: jest.fn(async () => {
+      return true
+    }),
```
note: the following 8 tests are currently using this new mock function.

```
git-directory-helper tests › cleans when clean true
git-directory-helper tests › checkout detach when not detached
git-directory-helper tests › does not checkout detach when already detached
git-directory-helper tests › does not clean when clean false
git-directory-helper tests › removes local branches
git-directory-helper tests › removes lock files
git-directory-helper tests › removes ancestor remote branch
git-directory-helper tests › removes descendant remote branch
```

Link- replied:
Yes @megamanics these tests cover the cleanup / folder deletion process. I want to see a test where the submodule status fails and we proceed to remove the folder and another one where it doesn't and we skip the cleanup.

megamanics replied:
@Link- sure will add additional 2 tests for the case and mock those responses.

megamanics replied:
🙇🏽 @sminnie for adding https://github.com/actions/checkout/pull/1196/commits/bb965fa266499cc8df6b104e7e448f6f6744cc61

sminnie replied:
@Link- - please review and let me know if the test cases are sufficient to merge?

jww3 replied:
New tests look good to me.

jww3 commented on src/git-command-manager.ts:
```diff
@@ -357,6 +358,12 @@ class GitCommandManager {
+    core.debug(output.stdout)
+    return output.exitCode === 0
+  }
+
```
for more info on the `git submodule` command, see:  https://git-scm.com/docs/git-submodule#Documentation/git-submodule.txt-status--cached--recursive--ltpathgt82308203

megamanics replied:
@jww3  This can be recreated by cloning https://github.com/test-SKi/test-bad-module-clone & using checkout action with any self-hosted runner to see the failure.

### Steps to recreate one scenario:
```
git checkout -b new
git clone git@github.com:test-SKi/submodule-repo.git
echo "recreate bad submodule issue" >>  README.md
git add .
git commit
git push --set-upstream origin new
```

joymixai commented:
ในวันที่ พ. 15 มี.ค. 2023 04:29 John Wesley Walker III <
***@***.***> เขียนว่า:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/git-command-manager.ts
> <https://github.com/actions/checkout/pull/1196#discussion_r1136230253>:
>
> > @@ -357,6 +358,12 @@ class GitCommandManager {
>      await this.execGit(args)
>    }
>
> +  async submoduleStatus(): Promise<boolean> {
> +    const output = await this.execGit(['submodule', 'status'], true)
> +    core.debug(output.stdout)
> +    return output.exitCode === 0
> +  }
> +
>
> for more info on the git submodule command, see:
> https://git-scm.com/docs/git-submodule#Documentation/git-submodule.txt-status--cached--recursive--ltpathgt82308203
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/actions/checkout/pull/1196#pullrequestreview-1340307118>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATEDT5K4L7B4BHS6EXTJ2HTW4DPLJANCNFSM6AAAAAAVNSLS64>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

sminnie committed:
Add 2  test cases for submodule status.
This commit adds tests to verify the behavior of the gitDirectoryHelper.prepareExistingDirectory() function when the submodule status is either true or false.
The test cleanWhenSubmoduleStatusIsFalse verifies that the function will clean the directory when the submodule status is false. The test sets up a mock implementation of git.submoduleStatus to always return false, writes a file to the repository, and then calls gitDirectoryHelper.prepareExistingDirectory(). The test verifies that the directory is cleaned and that git.tryClean() is called.
The test doesNotCleanWhenSubmoduleStatusIsTrue verifies that the function will not clean the directory when the submodule status is true. The test sets up a mock implementation of git.submoduleStatus to always return true, writes a file to the repository, and then calls gitDirectoryHelper.prepareExistingDirectory(). The test verifies that the directory is not cleaned, that the file and .git folder are present, and that git.tryClean() is called.
These tests ensure that the function behaves as expected based on the submodule status.

Ajaydip commented:
What is the status of this? should this PR be merged? We are experiencing [#993](https://github.com/actions/checkout/issues/933) as well.

kvanbere commented:
We would be very eager to have this merged if there is a chance that it might fix https://github.com/actions/checkout/issues/1148 , does anyone have a hypothesis whether this affects submodules only, or whether this change will also help runner recovery when the `.git` folder is corrupted ?

jww3 commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -42,7 +42,7 @@ jobs:
 
     - name: Initialize CodeQL
-      uses: github/codeql-action/init@v2
+      uses: github/codeql-action/init@v1
```
Why can't this remain `v2`?  Please clarify.

megamanics replied:
@jww3 This commit c8466ed95778604cdd7ea5a4875e87574070ed4c  change has been  merged from main to this branch.

Link- replied:
Hmm this shouldn't have happened, this has to stay `@v2`. In fact this entire file should not be part of this PR. I'll update it

Link- replied:
@megamanics - I'm not able to push to your fork i.e. this PR. We could do a rebase to fix this but this is simpler:
```
git checkout origin/main -- .github/workflows/codeql-analysis.yml
```
Could you please pull the workflow file from the main branch, commit and push it?

3Jz2 replied:
JamalZabetan@gmail.com

3Jz2 replied:
JamalZabetan@gmail.com

megamanics commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -55,4 +55,4 @@ jobs:
 
     - name: Perform CodeQL Analysis
-      uses: github/codeql-action/analyze@v2
+      uses: github/codeql-action/analyze@v1
```
```suggestion
      uses: github/codeql-action/analyze@v2
```

megamanics commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -42,7 +42,7 @@ jobs:
 
     - name: Initialize CodeQL
-      uses: github/codeql-action/init@v2
+      uses: github/codeql-action/init@v1
```
```suggestion
      uses: github/codeql-action/init@v2
```

megamanics reviewed (commented):
@Link- 6b842657bc27d31deeae7977f25f4f1ced77a653 reverted the version back to v2

megamanics reviewed (commented):
@Link- added commit so .github/workflows/codeql-analysis.yml is not part of this pr
https://github.com/actions/checkout/pull/1196/files

megamanics committed:
Codeql-Action Analyse revert v1 to v2

Link- reviewed (approved):
Thank you for this submission, I'm fine with this being merged 🚢

markjm commented:
Hi @megamanics / @Link- / @jww3 - would it be possible to cut a v3 release with these changes?
